### PR TITLE
MSE Dynamic Loss

### PR DESCRIPTION
**Pull request by jder**
Adds a new loss function to dynamically re-weight loss scaling to make each channel contribute equally to the loss.

### CODE REVIEW

File: configs/train_om4.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@ scheduler: true
 loss: mse
 finetune: false
 resume_ckpt_path: null
-inference_epochs: [-1]
+inference_epochs: []
```
Comment (jder) [inline configs/train_om4.yaml:14]: I made this work so we can just disable it altogether.

File: src/ocean_emulators/aggregator/validate/main.py [hunk 2]
```diff
@@ -31,7 +30,6 @@ def __init__(
         }
         self._aggregators = val_aggregators
         self.normalize = Normalize.get_instance()
-        self._loss_scaling = LossAggregator.get_instance().loss_scale
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/main.py:34]: LossAggregator had some loss scaling mechanisms that were never used. I deleted them and that let me delete the aggregator itself.

File: src/ocean_emulators/utils/logging.py [hunk 3]
```diff
@@ -215,7 +215,10 @@ def get_model_summary(model: torch.nn.Module, data: Mapping | None, debug: bool)
     logging.info(f"Number of parameters: {params}")
     depth = 10 if debug else 2
     # we pass verbose = 0 because we log the summary ourselves
-    logging.info(summary(model, input_data=[data], depth=depth, verbose=0))
+    if data is not None:
+        logging.info(summary(model, input_data=[data], depth=depth, verbose=0))
+    else:
+        logging.info(summary(model, depth=depth, verbose=0))
```
Comment (jder) [inline src/ocean_emulators/utils/logging.py:221]: This was a bug which broke eval. Made #317

File: src/ocean_emulators/train.py [hunk 4]
```diff
@@ -365,9 +383,9 @@ def __init__(self, cfg: TrainConfig) -> None:
         self.profiler = cfg.profiler.build(self.output_dir, self.device)
 
         assert self.tensor_map is not None
-        self.loss_aggregator = LossAggregator.init_instance()
 
-        self.init_inference_stores()
+        if self.inference_epochs:
+            self.init_inference_stores()
```
Comment (jder) [inline src/ocean_emulators/train.py:391]: Previously if you had no epoch to inference at, this would break.

File: src/ocean_emulators/config.py [hunk 5]
```diff
@@ -293,7 +293,13 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
 TrainBackendConfig = Literal["cpu", "cuda", "nccl", "auto"]
 LossType = Literal[
-    "mse", "mse_diff_weighted", "mse_cos_weighted", "mse_residual_scaled", "mse_mae"
+    "mse",
+    "mse_diff_weighted",
+    "mse_cos_weighted",
+    "mse_residual_scaled",
+    "mse_mae",
+    "mse_dynamic",
+    "mse_dynamic_no_limit",
```
Comment (alxmrs) [inline src/ocean_emulators/config.py:302]: I was hoping for "no cap"

Comment (alxmrs) [reply src/ocean_emulators/config.py:302]: (not a serious suggestion)

File: src/ocean_emulators/train.py [hunk 6]
```diff
@@ -72,6 +78,7 @@
     handle_warnings,
 )
 from ocean_emulators.utils.loss import (
```
Comment (alxmrs) [inline src/ocean_emulators/train.py:81]: What do you think about promoting `utils.loss` into `ocean_emulators.loss`?

File: src/ocean_emulators/train.py [hunk 7]
```diff
@@ -238,6 +245,7 @@ def __init__(self, cfg: TrainConfig) -> None:
         self.nets_dir = cfg.experiment.nets_dir
         self.network = cfg.experiment.network
 
+        self.loss_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor]
         # Loss function
         if cfg.loss == "mse":
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: 🐑  optional nit: this could be converted to a `match` expression. Though, I totally understand if you'd rather not, say, to minimize the diff.

File: src/ocean_emulators/utils/loss.py [hunk 8]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
     return combined.mean(dim=(0, 2, 3))
+
+
+N_WINDOW = 25
+
+
+class MseDynamic:
+    """A loss function that scales each channel contribute equally to the loss.
+
+    See: https://m2lines.slack.com/files/U086XUETAJ0/F0952H4G3GF/dynamic_loss_weighting.pptx
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: How about we check this into an `assets/` or `docs/` folder?

File: src/ocean_emulators/utils/loss.py [hunk 9]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
     mae = F.l1_loss(pred, target, reduction="none")
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
+
+
+N_WINDOW = 25
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: Could this live in the `constants.py` file?

File: src/ocean_emulators/utils/loss.py [hunk 10]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+        )
+        if should_limit:
+            vars: Float[torch.Tensor, " var"] = stds.pow(2)
+            self.limits: Float[torch.Tensor, " var"] | None = 1.0 / vars
+        else:
+            self.limits = None
+
+    def __call__(
+        self,
+        pred: Float[torch.Tensor, "batch hist*var lat lon"],
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:107]: Nice use of computed typing with `hist*var`

File: src/ocean_emulators/utils/loss.py [hunk 11]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+        target: Float[torch.Tensor, "batch hist*var lat lon"],
+    ) -> Float[torch.Tensor, " hist*var"]:
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:109]: What is the function of the space prefix? I believe that is intentional, but, what is it for?

File: src/ocean_emulators/utils/loss.py [hunk 12]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+        loss_with_history_channels: Float[torch.Tensor, " hist*var"] = decomposed_mse(
+            pred, target, self.wet
+        )
+        scaled_loss_including_history_dimension: Float[torch.Tensor, "hist var"] = (
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:113]: Happy about these jaxtyping annotations, makes it much easier to follow.

File: src/ocean_emulators/utils/loss.py [hunk 13]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
     mae = F.l1_loss(pred, target, reduction="none")
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
+
+
+N_WINDOW = 25
+
+
+class MseDynamic:
+    """A loss function that scales each channel contribute equally to the loss.
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: I'm a little confused by this sentence, is this what you mean?

```suggestion
    """A loss function that scales each channel contribution equally to the loss.
```

Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: I think additional description of what this does (and why it might be wanted) would be helpful.

File: src/ocean_emulators/utils/loss.py [hunk 14]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+            loss_with_history_channels.reshape(self.per_channel_scale.shape[0], -1)
+            * self.per_channel_scale.unsqueeze(1)
+        )
+        return scaled_loss_including_history_dimension.reshape(-1)
+
+    def update(
+        self,
+        pred: Float[torch.Tensor, "batch hist*var lat lon"],
+        target: Float[torch.Tensor, "batch hist*var lat lon"],
+    ):
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: I think this method could benefit from a docstring (or maybe a docstring and more descriptive name).

Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]:
```suggestion
    ) -> None:
```

File: src/ocean_emulators/utils/loss.py [hunk 15]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+        mse_loss = decomposed_mse(pred, target, self.wet)
+        mse_loss = torch.where(mse_loss == 0, 1e-8, mse_loss)
+        new_target_weights_with_history: Float[torch.Tensor, " hist*var"] = (
+            1.0 / mse_loss
+        )
+        # Reshape from channels * history to channels by averaging
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: :sheep: I was a little confused until I parsed this more in detail

```suggestion
        # Reshape from channels * history to channels by averaging (along the `hist` dimension)
```

File: src/ocean_emulators/utils/loss.py (hunk 9 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: After reading how it's used in the class below, I'd be happy to see this as a static class member, too.

Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: Furthermore, I think this constant could benefit by having a doctring, if only to explain the dimensions of the window. In my understanding, it's approximately timesteps or a window of history looking back? 

A modeling question I have: If I understand the units correctly, do we want a single look back window per every variable type? Put another way: would it ever make sense to use different window sizes to address differing time scales? To answer my own question (to check my understanding): the window can be fixed because we take aggregate statistics over the window, and the variances across each variable will naturally address the difference of each time scale, applied per each time scale.

File: src/ocean_emulators/utils/loss.py (hunk 14 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: This makes it clear to me that there is no return, as opposed to the return type merely being omitted.

File: src/ocean_emulators/utils/loss.py (hunk 9 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: Thinking aloud -- does this window function as a sort of "learning rate"?

File: src/ocean_emulators/utils/loss.py [hunk 16]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+
+        if get_world_size() > 1:
+            all_reduce_mean(new_target_weights)
+
+        self.per_channel_scale = (
+            self.per_channel_scale * (N_WINDOW - 1) + new_target_weights
+        ) / N_WINDOW
+
+    def loss_scale_per_channel(self) -> Float[torch.Tensor, " var"]:
+        return self.per_channel_scale
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py]: If we're exposing this to be read as a public method, then should the `per_channel_scale` member variable be private? On the other hand, why add this function, and not merely access the public member variable?

File: src/ocean_emulators/utils/loss.py [hunk 17]
```diff
@@ -67,3 +71,83 @@ def decomposed_mse_mae(
...
+
+    # new methods for saving and loading state
+    def state_dict(self) -> dict[str, torch.Tensor]:
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:151]: I could see these two methods being packaged into a protocol (happy to have in a later PR).

**Review by alxmrs — COMMENTED** [review]
Initial thoughts on the PR. I can't wait for dynamic weighting to land, this is quite promising.

File: src/ocean_emulators/train.py (hunk 6 above)
Comment (jder) [reply src/ocean_emulators/train.py:81]: I don't see much difference, what's the motivation?

File: src/ocean_emulators/utils/loss.py (hunk 8 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: My 2 cents is that it's fine to reference things in slack, GitHub, google drive, etc. If it was a markdown file we'd be editing over time I would feel differently but it's for reference and a binary blob.

File: src/ocean_emulators/utils/loss.py (hunk 11 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py:109]: It is necessary because of weird interaction with other linters: https://docs.kidger.site/jaxtyping/faq/#flake8-or-ruff-are-throwing-an-error

File: src/ocean_emulators/utils/loss.py (hunk 9 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: I moved this to a static class variable and added a doc string. For the modeling question, I would ask Adam if you want to discuss it.

File: src/ocean_emulators/utils/loss.py (hunk 13 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: Expanded a bit.

File: src/ocean_emulators/utils/loss.py (hunk 16 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: I was trying to be consistent with eg `state_dict` which is what Torch uses elsewhere and is a zero-arg function. I will add the underscore.

**Comment by jder** [comment]
@alxmrs thanks for the review! I think I addressed most of these, PTAL!

File: src/ocean_emulators/utils/loss.py (hunk 14 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: Added a docstring.

File: src/ocean_emulators/train.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py:81]: I see loss as a primary concern to the modeling rather than a utility.

File: src/ocean_emulators/utils/loss.py (hunk 8 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: My feeling is that it's often hard to remember to externalize links like this when we eventually go to open source this model + code. I like front-loading this sort of documentation. 

As a compromise: what about uploading the docs to Google Drive and pasting a Google Drive link (that we can then make open later on)?

File: src/ocean_emulators/utils/loss.py (hunk 9 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: Happy to discuss later. I'll think more about how it affects the modeling. In any case, it's an interesting approach and I'm glad it works well!

Thanks for making the other changes.

File: src/ocean_emulators/utils/loss.py (hunk 11 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py:109]: Thanks for the pointer! That makes sense.

File: src/ocean_emulators/utils/loss.py (hunk 13 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: Thanks!

File: src/ocean_emulators/train.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py:81]: Happy to consider this later; not blocking landing this PR.

File: src/ocean_emulators/utils/loss.py [hunk 18]
```diff
@@ -67,3 +71,88 @@ def decomposed_mse_mae(
     mae = F.l1_loss(pred, target, reduction="none")
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
+
+
+class MseDynamic:
+    """A loss function that scales each channel to contribute equally to the loss.
+
+    This uses a rolling estimate of the loss of each channel to scale each
+    channel's loss, discouraging the model from focusing on only a few channels.
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:80]: Thanks, this explanation is really helpful!

File: src/ocean_emulators/utils/loss.py [hunk 19]
```diff
@@ -67,3 +71,88 @@ def decomposed_mse_mae(
...
+
+    See: https://openathena.slack.com/archives/C08CYM42DT3/p1752275713570969
+    """
+
+    N_WINDOW = 25
+    """Rolling window size to average over. (~number of steps)"""
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:86]: nit: I think it's conventional to go in the other order.

```suggestion
    """Rolling window size to average over. (~number of steps)"""
    N_WINDOW = 25
```

**Review by alxmrs — APPROVED** [review]
Thanks for the changes, this PR LGTM! Excited to have this feature in main. 

(Sorry for posting the review comments one by one)

File: src/ocean_emulators/utils/loss.py (hunk 8 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: If we end up open sourcing this code (eg we may not if the experiments don't work out), I don't think we'd want to necessarily release that PPT as well. We're going to have to come back and review/remove/change when that happens. If we're developing in the open then of course this is easier, but for now I don't think we should spend time getting ready to open source each code change.

File: src/ocean_emulators/utils/loss.py (hunk 19 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py:86]: I agree it is weird, but the [relevant PEP](https://peps.python.org/pep-0257/#what-is-a-docstring) says:

> String literals occurring immediately after a simple assignment at the top level of a module, class, or __init__ method are called “attribute docstrings”.

And according to [various folks on SO](https://stackoverflow.com/questions/8820276/docstring-for-variable), this is what Sphinx, IntelliJ, VS Code, etc, support.

**Comment by jder** [comment]
Thanks for the re-review! I extracted out #326 and replied to your other comments. Planning to merge when green but happy to discuss if open threads afterwards.
